### ISSUE TITLE
questions and issues regarding training with nnUNetPlannerResEncL on BraTS2023

### ISSUE THREAD

**Issue by neuronflow** [MIC-DKFZ/nnUNet#2653]
I am trying to train nnUNet on BraTS 2023 data (1251 exams).
The idea is to train on all cases and then evaluate on a separate test set.

With the standard `3d_fullres` config, everything works without trouble.
However, nnUNet advertises I should use the residual encoders, so I try:

```
nnUNetv2_train 1337 3d_fullres all -p nnUNetPlannerResEncL -num_gpus 1
```

Notably, the training reaches pseudo dice 1 for some channels and .999 for others (can this flavor of nnUNet overfit the BraTS training set really so well?)

After training, I tried to run inference on our test set. Therefore, I used:
```
nnUNetv2_predict -i /input/imagesTs -o /output/predictions -d 1337 -f all -c 3d_fullres -p nnUNetResEncUNetLPlans 
```

For this to work, I had to manually copy some .json files. Am I using the wrong command here?

The resulting segmentations are terrible:
![image](https://github.com/user-attachments/assets/2b4a9adb-50ad-4121-99d2-4f904cbef96f)

**Comment by sten2lu** [comment]
Hi @neuronflow,

First off, it is not the intended behavior that json files need to be copied around. Could you please give me information which json files you had to copy manually and also into which places?

Secondly, the residual encoder nnU-Net can overfit onto the data much easier than the standard model.  

To get a better understanding what might be going on, please could you do the following:
1. Give me more details with regard to the dataset you use for testing the model. Is it also the BraTS dataset? And can you provide performance metrics s.a. mean dice/ dice per class etc?
2. Send me the progress.png of your training located inside the nnUNet_results folder.

Best regards,

Carsten

**Comment by neuronflow** [comment]
Dear @sten2lu,

I am testing on the non-public BraTS validation set. Similar distribution to the training set. Mean dice scores are very low, between 5-10% depending on the class, due to the heavy over-segmentation.

I copied the `plans.json`
<img width="367" alt="Screenshot 2024-12-30 at 16 04 54" src="https://github.com/user-attachments/assets/482dd70a-524e-4c96-bd8f-b81446571559" />


the progress:
![](vscode-remote://ssh-remote%2Bgiga/mnt/Drive4/florian/brats_modality_ablation_nnUnet/nnu_data/results/Dataset1337_BraTS2023_modality_ablation/nnUNetTrainer__nnUNetResEncUNetLPlans__3d_fullres/fold_all/progress.png)

**Comment by sten2lu** [comment]
Hi @neuronflow,

Sorry for the long response time and we are actively looking into it.
I could reproduce your error regarding the ResEnc Plans.

At the moment, I try to reproduce this Issue on the BraTS21 dataset and the BraTS23 dataset.
I will get back with you on Monday when the trainings are finished.

Best regards,
Carsten

**Comment by sten2lu** [comment]
Hi @neuronflow,
I got results on the BraTS21 dataset and have a DICE score of ~0.92.
Also the pseudo dice does not converge to 1.

Could you please send me the dataset.json alongside how you preprocessed the BraTS23 dataset?
So that we can reproduce this issue?

**Comment by neuronflow** [comment]
Sorry, for the late response.

> I could reproduce your error regarding the ResEnc Plans.

I guess you mean the renaming issue with this?

my preprocessing script for training data:
```
import shutil

from batchgenerators.utilities.file_and_folder_operations import *
from nnunetv2.dataset_conversion.generate_dataset_json import generate_dataset_json
from nnunetv2.paths import nnUNet_raw

if __name__ == "__main__":
    brats_data_dir = "/mnt/Drive4/florian/brats_modality_ablation_nnUnet/data/vanilla/ASNR-MICCAI-BraTS2023-GLI-Challenge-TrainingData"

    task_id = 1337
    task_name = "BraTS2023_modality_ablation"

    foldername = "Dataset%03.0d_%s" % (task_id, task_name)

    # setting up nnU-Net folders
    out_base = join(nnUNet_raw, foldername)
    imagestr = join(out_base, "imagesTr")
    labelstr = join(out_base, "labelsTr")
    maybe_mkdir_p(imagestr)
    maybe_mkdir_p(labelstr)

    case_ids = subdirs(brats_data_dir, prefix="BraTS", join=False)

    for c in case_ids:
        shutil.copy(
            join(brats_data_dir, c, c + "-t1n.nii.gz"),
            join(imagestr, c + "_0000.nii.gz"),
        )
        shutil.copy(
            join(brats_data_dir, c, c + "-t1c.nii.gz"),
            join(imagestr, c + "_0001.nii.gz"),
        )
        shutil.copy(
            join(brats_data_dir, c, c + "-t2w.nii.gz"),
            join(imagestr, c + "_0002.nii.gz"),
        )
        shutil.copy(
            join(brats_data_dir, c, c + "-t2f.nii.gz"),
            join(imagestr, c + "_0003.nii.gz"),
        )
        shutil.copy(
            join(brats_data_dir, c, c + "-seg.nii.gz"),
            join(imagestr, c + "_0003.nii.gz"),
        )

        shutil.copy(
            join(brats_data_dir, c, c + "-seg.nii.gz"),
            join(labelstr, c + ".nii.gz"),
        )

    generate_dataset_json(
        out_base,
        channel_names={
            0: "T1",
            1: "T1ce",
            2: "T2",
            3: "Flair",
        },
        labels={
            "background": 0,
            "whole tumor": (1, 2, 3),
            "tumor core": (2, 3),
            "enhancing tumor": (3,),
        },
        num_training_cases=len(case_ids),
        file_ending=".nii.gz",
        regions_class_order=(1, 2, 3),
        license="see https://www.synapse.org/#!Synapse:syn25829067/wiki/610863",
        reference="see https://www.synapse.org/#!Synapse:syn25829067/wiki/610863",
        dataset_release="1.0",
    )

```

Attached are all seemingly relevant jsons I could find.

[dataset_fingerprint.json](https://github.com/user-attachments/files/18645272/dataset_fingerprint.json)
[dataset.json](https://github.com/user-attachments/files/18645273/dataset.json)
[nnUNetPlannerResEncL.json](https://github.com/user-attachments/files/18645271/nnUNetPlannerResEncL.json)
[nnUNetResEncUNetLPlans.json](https://github.com/user-attachments/files/18645270/nnUNetResEncUNetLPlans.json)

**Comment by kalazus** [comment]
>         shutil.copy(
>             join(brats_data_dir, c, c + "-t2f.nii.gz"),
>             join(imagestr, c + "_0003.nii.gz"),
>         )
>         shutil.copy(
>             join(brats_data_dir, c, c + "-seg.nii.gz"),
>             join(imagestr, c + "_0003.nii.gz"),
>         )
> 
>         shutil.copy(
>             join(brats_data_dir, c, c + "-seg.nii.gz"),
>             join(labelstr, c + ".nii.gz"),
>         )

Second to last `shutil.copy` rewrites Flair modality with segmentation mask

**Comment by lucas-santana** [comment]
@neuronflow 
In BraTS 2023, the labels are:

* 1 for NCR
* 2 for ED
* 3 for ET
* 0 for everything else

So the regions are defined as:

* WT = NCR (1) + ED (2) + ET (3)
* TC = NCR (1) + ET (3)
* ET = ET (3)

In your preprocessing script, you define at the end:

```python
labels = {
    "background": 0,
    "whole tumor": (1, 2, 3),
    "tumor core": (2, 3),
    "enhancing tumor": (3,),
}
```

Had you already performed the conversion of 1 to 2 and 2 to 1, as done in lines 24 and 25 of the nnUNetv2 dataset conversion script for BraTS21 ([https://github.com/MIC-DKFZ/nnUNet/blob/master/nnunetv2/dataset_conversion/Dataset137_BraTS21.py](https://github.com/MIC-DKFZ/nnUNet/blob/master/nnunetv2/dataset_conversion/Dataset137_BraTS21.py))?
I didn’t see that conversion in your script.
